csi/src/node.rs

```diff
@@ -298,31 +296,74 @@ impl node_server::Node for Node {
             ));
         }
 
-        // target path will have been created previously in node_publish_volume
-        // and is one of
-        //  1. a directory for filesystem volumes ,
-        //  2. a block special file for block volumes.
-        //
-        // If it does not exist, then a previously unpublish request has
-        // succeeded.
-        let target_path = Path::new(&msg.target_path);
-        if target_path.exists() {
-            if target_path.is_dir() {
-                unpublish_fs_volume(&msg)?;
-            } else {
-                if target_path.is_file() {
-                    return Err(Status::new(
-                        Code::Unknown,
-                        format!(
-                            "Failed to unpublish volume {}: {} is a file.",
-                            &msg.volume_id, &msg.target_path
-                        ),
+        if mount::find_mount(None, Some(&msg.target_path)).is_some() {
+            debug!("Unmounting {}", msg.target_path);
+            if let Err(error) = mount::bind_unmount(&msg.target_path) {
+                return Err(failure!(
+                    Code::Internal,
+                    "Failed to unpublish volume {}: failed to unmount {}: {}",
+                    msg.volume_id,
+                    msg.target_path,
+                    error
+                ));
+            }
+        }
+
+        let mount_point = Path::new(&msg.target_path);
+        match fs::metadata(mount_point) {
+            Ok(metadata) => {
```

Comment:
nit: might be able to reduce if/else nesting by using:
```rust
Ok(metadata) if metadata.is_dir() => ...
Ok(metadata) if metadata.is_file() => ...
```

Reply:
I can change this if you feel strongly, but I think it doesn't really make much difference one way or the other.

Reply:
no not really, it's just personal preference nit :)